README.md
# <p align="center"> MKlon2v5
# <p align="center">Реализация конечного автомата в проекте зарядного устройства.
## <p align="center">Модульное зарядное устройство на ESP-WROOM-32D - проект Arduino.
#### <p align="right">редакция от 6 декабря 2022 года
![](https://github.com/olmoro/Mklon2v5/blob/main/Documents/Img/Mklon2v5.jpg) Базовая версия.
### Это проект для тех, кто наигрался с прошивками и профилями, а дух исследователя подгоняет продолжать поиски новых технических решений вплоть до реализации собственных алгоритмов. В этом случае имеет значение порог вхождения, то есть тот уровень знаний и навыков, необходимых для реализации задуманного. Сможете написать программу для елочной гирлянды на Arduino? имеете навыки пайки печатных плат с элементами SMD? имеете ЗУ типа Кулон, который вам уже скучен урезанным функционалом, но выводить его из строя улучшениями как-то не хочется? - тогда это для вас. Воплотив свою мечту в одночасье заменяете начинку корпуса. По сути предлагается своего рода глубокий "рестайлинг" Кулона, то есть вдохнуть больше жизни в прибор, убрав ненужное и добавив актуальное. Точно знаете что и как хотите реализовать - тогда вперёд.  
1. [Как это работает](#mklon)
2. [Аппаратный аспект](#hardware)
    - [Платформа](#hard_platform)
    - [Модули](#modules)
        - [Управление](#control)
        - [Индикация](#lcd)
        - [Измерение](#measure)
        - [Регулирование](#adjusting)
        - [Питание](#power)
        - [Таймер](#timer)
    - [Клавиатура](#key)
    - [Датчики](#sensor)
    - [Диагностика](#diagnostics)
    - [Подключение батареи](#connecting)
    - [Система охлаждения](#cool)
    - [Разряд](#discharge)
    - [Термоконтроль](#thermal)
    - [Защита](#protection)
        - [Переполюсовка](#reverse_polarity)
        - [Перегрузка](#overload)
        - [Искрогашение](#spark)
    - [По мелочи](#small)

3. [Программный аспект](#soft)
    - [Платформа](#soft_platform)
    - [Дисплей](#128x160)
    - [Клавиатура](#button)
    - [ПИД-регулятор](#pid)
    - [Сеть](#connect)
    - [Протокол обмена](#wake)
    - [Команды](#commands)
    - [АЦП](#adc)
    - [Утилиты](#tools)
4. [Технические характеристики](#specifications)
5. [FSM](#fsm)
6. [Полезные ссылки](#useful_links)
7. [About Me](#about)
***
## <p align="center"> Как это работает.
В зарядном устройстве пользовательский интерфейс реализован на модуле с ESP32 (ESP), а управление силовым блоком производится по последовательному каналу. На ESP обслуживается клавиатура, дисплей, RGB индикатор, зуммер, система контроля питания, датчики температуры, система охлаждения, интерфейс драйвера силового блока, часики бортового времени, USB, WiFi, BT. В памяти ESP хранятся заводские и пользовательские настройки. Оператору предоставляется возможность выбрать посредством меню требуемый режим работы, задать параметры и запустить. Настройки беспроводной сети позволяют в реальном времени на удаленном компьютере не только наблюдать в виде графиков процесс: напряжение, ток, баланс заряда/разряда, температуру, работу системы охлаждения, но и управлять им. Программное обеспечение ESP построено таким образом, что позволяет даже непрофессиональному программисту реализовать свой собственный алгоритм, воспользовавшись методикой из репозитория FSM (собственно ради этого и задумывался проект).

Драйвер силового блока (DRV) как таковой может отсутствовать - свободных ресурсов ESP достаточно, а INA226, ADS1115 можно подключить к последовательному каналу, установив свой модуль. Но здесь представлен более интересный вариант, когда весь, или почти весь реалтайм реализован на отдельном модуле SAMD21 MINI (D21). Связь c ESP осуществляется по асинхронному каналу. Модуль D21 получает команду, по которой включаются соответствующие ресурсы: АЦП тока и напряжения, ШИМ-генератор, ЦАП управления разрядом, коммутатор выхода и др. Автоматически поддерживается устойчивость силового преобразователя при работе на малую нагрузку или на холостом ходе, отключение при перегрузках. Поддерживается режим быстрого разряда накопительных конденсаторов через нагрузку для разряда, минуя датчик тока, дабы не искажать подсчет ампер-часов. Когда возможно допускается регулировка параметров "на лету", иначе D21 переводит силовую часть в безопасный режим с выдачей сообщения управляющему контроллеру. D21 поддерживает управление зарядом и разрядом как командами, так и с помощью собственного ПИД-регулятора. Настройки коэффициентов для поддержания тока и напряжения могут быть разные. Выбор ПИД-регулирования имеет 4 режима: отключено, по току, по напряжению и с автовыбором по типу реализованного в TL494. Измерения напряжения на батарее производятся по четырехпроводной схеме в диапазоне от -2 до +18 вольт. Ток разряда измеряется на том же шунте, что и ток заряда от -10 до +10 ампер. Параметры фильтрации измерений задаются командами.

Питание прибора производится от внешнего AC/DC преобразователя на 18...20 вольт 6...9 ампер. Модули защищены самовосстанавливающимся предохранителем. В качестве "последнего барьера" на выходе силового преобразователя защитный диод и предохранитель - но это так, исторически...

![](https://github.com/olmoro/Mklon2v5/blob/main/Documents/Img/МКлон(прототип).jpg) Прототип. Испытание под активной нагрузкой. 

***
## <p align="center"> Аппаратный аспект <a name="hardware"></a>
### <p align="center">Платформа <a name="hard_platform"></a>
Arduino как аппаратная платформа предполагает модульный принцип построения системы (читай: многое уже сделано за нас, а если проблема решается за деньги, то это не проблема, а расходы)) - с этого и начнем... 
### <p align="center">Модули <a name="modules"></a>
#### Управление <a name="control"></a>
Управляющий модуль [ESP32-WROOM-32D](https://aliexpress.ru/item/32995519416.html?spm=a2g2w.orderdetails.0.0.1c024aa6EZlC3f&sku_id=10000000045993301&_ga=2.246283140.1610960066.1667852261-603987165.1661065985) на 4 или 16МБ распаян на плату. Питание модуля - от 18-вольтового источника через DC/DC преобразователь до 5.5 вольт, "где смешивается" схемой на диодах с напряжением, снимаемым с USB-разъёма, и далее на LM1117 понижается до требуемых 3.3 вольт. Такая схема вкупе с параметрами цепи сброса обеспечивает надёжный старт при подаче питания. Для программирования модуля используется упомянутый USB-интерфейс на CP2104. 
#### Индикация <a name="lcd"></a>
Модуль дисплея - полноцветный [TFT](https://aliexpress.ru/item/32996979276.html?spm=a2g2w.orderdetails.0.0.f9664aa6FFnV7N&sku_id=67278796295&_ga=2.258282378.1610960066.1667852261-603987165.1661065985) с подсветкой. На плате предусмотрена возможность подключения как 128х128, так и 128х160 варианта. Подключен по схеме SPI-3. На плате версии имеется дополнительный разъем для установки "кулоновского" ЖКИ типа S69717-BF016MC (128x128), но без подключения требуемой для него внешней платы подсветки. Сообществу почитателей Arduino вряд ли понравится, что публичной библиотеки на него нет.
#### Измерение <a name="measure"></a>
Модуль измерения [WeMos SAMD21 M0 Mini](https://aliexpress.ru/item/4001074203011.html?_ga=2.268755377.1168692661.1643047564-769654542.1642920280&sku_id=10000014182967685&spm=a2g39.orderlist.0.0.60754aa6RxSv8u) съемный, устанавливается на цанговые [разъемы](https://www.chipdip.ru/product0/8002610753) или [панель](https://www.chipdip.ru/product/trl-28). Идущие в комплекте штыревые разъемы лучше не применять ввиду их ненадежности. Модуль размещен "в тени" обдувающего потока и имеет доступ к своему USB-MICRO с тыльной стороны прибора. Для питания модуля используется гальванически развязанный [источник](https://aliexpress.ru/item/1005003568781579.html?_ga=2.165478714.428003227.1651560775-769654542.1642920280&sku_id=12000026320337806&spm=a2g39.orderlist.0.0.50684aa6oqK3eu) на 5 вольт и мощностью 1 ватт. Связь с управляющим контроллером по асинхронному интерфейсу 230400 бод.
#### Регулирование <a name="adjusting"></a>
Силовой преобразователь [XL4016 DC-DC Max 9A 300W](https://aliexpress.ru/item/32661166685.html?_ga=2.264151500.1168692661.1643047564-769654542.1642920280&sku_id=12000020944321893&spm=a2g39.orderlist.0.0.60754aa6RxSv8u) был демонтирован и частично перенесен на плату. Драйвер ключа заменен на MIC4420 с сохранением рабочей частоты 190кГц.
#### Питание <a name="power"></a>
Модуль питания AC/DC на 6А внешний, например, [WX-DC2406](https://aliexpress.ru/item/1005001714079538.html?_ga=2.268223694.1168692661.1643047564-769654542.1642920280&sku_id=12000018559376968&spm=a2g39.orderlist.0.0.60754aa6RxSv8u), настроенный на выходное напряжение 18 вольт. Или аналогичный на 9А (не проверялось). Как вариант может использоваться ноутбучный [источник питания](https://market.yandex.ru/product--blok-pitaniia-zariadka-dlia-noutbuka-asus-19v-6-32a-5-5-2-5-setevaia-adp-120rh/1734845969?cpc=bcjhtLvFsAaOy2Y0EAbUIRvKu3N5uSGibT0yC8CfX7qp3RRzn_rff3QEG0-vM6r_ZKvIfzGDGmySv3K-t1ayx6ZddH5EJmUSNwWh55xaryzd0VnnOPKlpaGhPmwbtu7LhoPJAQeR6_Byu1-WVCNUIS1UXxPNSBOymYskzyHlVkIgZ9DcLLzQeE4s890wWZX36JCk2veC-4o%2C&from-show-uid=16690006279223628928500002&sku=101611095356&do-waremd5=AGT0s5TgN0KGJ9slrCg1gw&sponsored=1&cpa=1) 19V 6.32A. Дороже, зато не надо мудрить с "окорпусовыванием" WX. Вместо встроенной электронной нагрузки предлагается возможность подключения в режиме разряда нагрузки пользователя. Разъем от автомобильного "прикуривателя" на задней стенке прибора позволит подключить многое из арсенала автолюбителя - от лампы подсветки до USB-зарядки или DС/AC на 230 вольт выходного напряжения. Подкачать колеса вряд ли получится, но побриться - вполне, и электричество не растратится впустую, и "зеленые" будут довольны.  
#### Таймер<a name="timer"></a>
Модуль [прецизионных часов](https://aliexpress.ru/item/32824174999.html?sku_id=64909286317&spm=.list.1.6cc1587eXZyGas) на DS3231 с батарейкой, лучше литиевой с возможностью подзаряда. Или же специальный модуль таймера со слотом для карты памяти формата MICRO. ![](https://github.com/olmoro/Mklon2/blob/main/Documents/Img/timer_sd.jpg)
### <p align="center">Клавиатура <a name="key"></a>
Схема 5-кнопочной клавиатуры при нажатии выдает соответствующий уровень напряжения, причем резисторы подобраны таким образом, что опорное напряжение делится на равные доли. Выглядит довольно брутально, но кнопки могут быть и поизящнее.
### <p align="center">Датчики <a name="sensor"></a>
Датчики измерителей напряжения и тока включены для дифференциальных измерений со смещением в разумных пределах общего провода в "минус", что позволяюет измерять напряжение и ток обеих полярностей.
### <p align="center">Диагностика <a name="diagnostics"></a>
На плате имеются два разъёма для беспроводного подключения 8-канального логического [анализатора](https://aliexpress.ru/item/1005004783286574.html?sku_id=12000030470334955&spm=a2g2w.productlist.list.6.4d1b5c5e8YmNAb) к некоторым цепям модулей ESP32 и SAMD21, что может быть весьма полезно при отладке устройства и разработки целевого ПО.
### <p align="center">Подключение батареи <a name="connecting"></a>
Батарея подключается по четырехпроводной схеме, понижающей в значительной степени требования к качеству соединительных проводов, а также исключающих их калибровку. Но для этого потребуются соответствующие ["крокодилы"](https://aliexpress.ru/item/4000593884517.html?spm=a2g2w.orderdetails.0.0.777a4aa6ri0LqM&sku_id=12000020938249496&_ga=2.53804840.1610960066.1667852261-603987165.1661065985).
### <p align="center">Система охлаждения <a name="cool"></a>
Силовые элементы (все в TO-247), расположенные на одном общем радиаторе с термодатчиком NTC 10ком, обдуваются 60х15 мм вентилятором на 12 вольт. Предусмотрена обратная связь по третьему проводу. Преобразование ШИМ выхода ESP в управляющее скоростью напряжение производится схемой на MIC4420. Оригинальное крепление двух TO-247 одним винтом на разных сторонах плоского радиатора объясняется тем, что они не работают одновременно. Что вряд ли понравится почитателям рестайлинга. Радиатор заимствован с Кулона-912 и расположен точно по оси воздушного потока от вентилятора, что несколько их утешит... Для вентилятора  60 мм предусмотрена ниша глубиной 15 мм.
### <p align="center">Разряд <a name="discharge"></a>
Схема разряда предусматривает возможность подключения внешней нагрузки для облегчения теплового режима прибора. На задней стенке прибора можно установить [разъём](https://aliexpress.ru/item/1005004687429739.html?sku_id=12000030101513797&spm=.list.5.371b2aa3ztIT9A). Опасения насчет того, что сопротивление кастомной нагрузки непрогнозируемо, снимается, так как в отличие от кулоновских схем здесь ток разряда не только задается, но и измеряется, а потому подсчет слитых ампер-часов более корректен. К тому же появляется возможность подсчитать внутреннее сопротивление батареи.
### <p align="center">Термоконтроль <a name="thermal"></a>
К прибору может быть подключена сеть из разумного числа термодатчиков типа DS1820 для контроля, например, температуры побаночно. Интерфейс - DS2482-100S. Разъем RJ11-6 может быть вынесен на тыльную или боковую стенку прибора, как и карта памяти).
### <p align="center">Защита <a name="protection"></a>
#### Переполюсовка <a name="reverse_polarity"></a>
Защита от переполюсовки батареи - есть.
Защита от переполюсовки питания - отсутствует, так как разъемы питания полярные.
#### Перегрузка <a name="overload"></a>
Защита от перегрузки по току заряда - есть.
Защита от перегрузки по току разряда - есть.
#### Искрогашение <a name="spark"></a>
Функция искрогашения при подключении батареи - есть.
### <p align="center">По мелочи <a name="small"></a>
Зуммер - активный. Светодиод RGB 5 мм. Кнопка аппаратного сброса. Печатная плата в габарите Кулона-912, 113х128мм двусторонняя. 
***
## <p align="center">Программный аспект <a name="soft"></a>
### <p align="center">Платформа<a name="soft_platform"></a>
Arduino как программная платформа предполагает широкое использование сторонних разработок в виде библиотек, как правило на безвозмездной основе. (Цена вопроса потому - время на поиск/подбор подходящей библиотеки).
### <p align="center">Дисплей<a name="128x160"></a>
Driver IC:ST7735S поддерживается многими библиотеками, в том числе и TFT_eSPI by Bodmer. К достоинству следует отнести подробную инструкцию по настройке для выбранного подключения. К сожалению кириллицы нет (а может быть уже есть).
### <p align="center">Клавиатура<a name="button"></a>
Для схемы с делителем напряжения и аналоговым входом применена библиотека ... точнее [публикация](http://samopal.pro/arduino-button-2/). Возможности вполне устраивают: различаются короткие и длинные нажатия, автоповтор. Тайминги задаются по каждой кнопке свои.
### <p align="center">Пид-регулятор<a name="pid"></a>
Пид-регулятор, реализованный в DRV, [FastPID](https://github.com/mike-matera/FastPID/tree/master/examples/VoltageRegulator). Данные передаются целочисленными и предвычисленными для сокращения времени обработки драйвером силовой части.

 ![](https://github.com/olmoro/MKlon2v5/blob/main/Documents/Img/PID_Compensation_Animated.gif)

Влияние изменения параметров PID (Kp,Ki,Kd) на переходную характеристику системы. 
Подбор коэффициентов - шаманство ещё то, тем более если в наличии только 5-кнопочная клавиатура. Надежда на управление по беспроводной сети.
### <p align="center">Сеть<a name="connect"></a>
На этапе разработки и отладки применен пакет [AutoConnect](https://github.com/Hieromon/AutoConnect) Заимствован "как есть", добавлен график тока и напряжения с автоматическим выбором диапазонов. 
### <p align="center">Протокол обмена<a name="wake"></a>
Выбор протокола WAKE объясняется его минимальной избыточностью. Универсальные по этому параметру существенно проигрывают.
### <p align="center">Команды<a name="commands"></a>
Список команд насчитывает несколько десятков, в основном исследовательских. А в мирной жизни нужны не более десятка-двух - такие как считать напряжение, ток и состояние (причем в физических единицах или "попугаях" АЦП), включить, отключить, задать параметры и т.п. 
### <p align="center">АЦП <a name="adc"></a>
Слабенький собственный АЦП ESP32 мало пригоден для измерений, но с контролем клавиатуры и отслеживанием питающего напряжения справляется вполне.
### <p align="center">Утилиты<a name="tools"></a>
Список системных утилит - это по сути набор инструментов, которые используются при реализации пользовательских алгоритмов. Особенность их состоит в том, что в них заложена проверка, не допускающая разрушения прибора, ограничивая намерения разработчика алгоритма. Если пользователь хорошо представляет реализуемый алгоритм, то написать программу не сложнее, чем для ёлочной гирлянды, тем более, что один из реализованных режимов - шаблон с подробными пояснениями. Но уметь программировать в среде Arduino надо уверенно. На профессиональном уровне всё для этого уже решено.
***
## <p align="center">Технические характеристики MKlon2v5 (вариант D21)<a name="specifications"></a>
### <p align="center">Режим заряда
    - Зарядный ток Iз, А ......................... 0.05 – 10.0 ........ ±(0.005 Iз + 0.05)*
    - Шаг установки зарядного тока, А ............ 0.01 
    - Зарядное напряжение U, В ................... 1.0 – 18.0 ......... ±(0.005 U + 0.05)*
    - Шаг установки зарядного напряжения, В	...... 0.01
### <p align="center">Режим разряда
    - Разрядный ток Iр, А ........................ 0.05 – 10.0 ........ ±(0.005 Iр + 0.05)*
    - Шаг установки разрядного тока, А ........... 0.01
    - Максимальная рассеиваемая мощность, Вт ..... не менее 40
    - то же с внешней нагрузкой, Вт .............. не более 120
    - Пределы измерения сопротивления R, мОм ..... 1 - 6000 ........... ±(0.05 R)*
 \* - уточняются по результатам испытаний
### <p align="center">и наконец
    - Питание от внешнего AC/DC .................. 20 В, 6 А или ноутбучный 19 В (4.74 ... 9.5 А)
    - Максимальная потребляемая мощность ......... 200 В·А
    - Интерфейс .................................. 5 кнопок, USB, WiFi, BT, 1-WIRE
    - Поддержка датчиков температуры ............. до 6 штук
    - Индикация .................................. RGB 1.8" 128 X 160 (Driver IC:ST7735S)
    - Диагностика ................................ 2 разъема для логического анализатора
    - Встроенные часы реального времени .......... с автономным источником
    - Micro-SD карта ............................. предусмотрена, 16 Гбайт
    - Защита по выходу ........................... переполюсовка, перегрузка по току, КЗ
    - Степень защиты корпуса ..................... IP20
    - Климатические факторы: ..................... ГОСТ 15150, УХЛ 4
	    - диапазон рабочих температур ............ от 5 до 40 °С
	    - относительная влажность воздуха ........ 80 % при температуре 25 °С
	    - атмосферное давление ................... От 84 до 106.7 кПа
    - Условия хранения:
	    - диапазон температур .................... От -25 до 55 °С
	    - относительная влажность воздуха ........ 95 % при температуре 30°С
    - Габаритные размеры корпуса ................. 153 х 85 х 215 мм
    - Масса без комплекта кабелей и внешнего ИП .. не более 0.8 кг
    - Средний срок службы ........................ 10 лет
    - Средняя наработка на отказ ................. 25000 ч
    - Цена ....................................... определяется самостоятельной сборкой
***
# <p align="center">FSM <a name="fsm"></a>
## <p align="center"> Finite State Machine или finite-state machine или конечный автомат, кому как нравится. 
"Это до предела упрощенная модель компьютера, имеющая конечное число состояний, которая жертвует всеми особенностями компьютеров такие как ОЗУ, постоянная память, устройства ввода-вывода и процессорными ядрами в обмен на простоту понимания, удобство рассуждения и легкость программной или аппаратной реализации." Поверим пока на слово.

### <p align="center"> Об одном способе реализации конечного автомата.
Идея состоит в том, что реализация каждого режима работы устройства должна производиться независимо от иных режимов, но использовать некий общий инструментарий, обеспечивающий безаварийный доступ к ресурсам прибора. Как говорил великий Х.Д.Милс: "Нахождение глубинной простоты в запутанном клубке сущностей - это и есть творчество в программировании." - не дословно, но близко к источнику. В свое время мне понадобилось не более двух месяцев, чтобы подавить в себе естественное отторжение и привести проект к виду, с которым комфортно и, главное, безопасно работать. Ну-ну.

### <p align="center"> Конечный автомат, он же Finite State Mashine (FSM).
Любой или почти любой процесс можно представить в виде последовательности шагов от одного состояния к другому. Разработчик мало чего стоит, если не представляет себе все возможные состояния, число которых конечно - именно поэтому автомат и называют "конечным". Вдохновением можно запастись [здесь](https://chipenable.ru/index.php/programming-avr/item/90-realizatsiya-konechnogo-avtomata-state-machine.html). Способ несомненно хорош, Но вот синхронизировать данные двух структур... хвост может начать вилять собакой.

Как же будет выглядеть реализация конкретного режима, например простого заряда аккумуляторной батареи? Графически режим такого заряда может быть представлен так: идите от жирной точки вверху, в цветных бланках короткие примечания к состояниям.
![Заряд CC/CV](https://github.com/olmoro/MKlon2v5/blob/main/Documents/FSM/CC_CV_fsm.jpg)
Вырисовывается не так уж и много состояний - чуть более десятка. Слева - состояния ввода параметров с возможностью отказа от дальнейшего ввода и перехода на исполнение или выход. Справа - последовательность шагов заряда - подъем тока, удержание тока и удержание напряжения. В центре - состояние отложенного старта. Исходим из того, что этот режим заряда (CC/CV) у нас "в печёнках", у кого нет - представьте, что это алгоритм ёлочной гирлянды. 
Оформим каждое состояние соответствующим объявлением класса в общем поле имен CcCvFsm. Здесь же можно определить необходимые константы, свойственные для этого режима, чтобы всегда были "под рукой" в структурах, а в классах ничто не мешает определить свои локальные константы и переменные как private:

##### файл cccvfsm.h
```c++
#ifndef _CCCVFSM_H_
#define _CCCVFSM_H_

#include "mstate.h"

namespace CcCvFsm    // Поле имен для режима простого заряда
{
  struct MChConsts
  {
      // Пределы регулирования
      static constexpr float i_l =  0.2f;
      static constexpr float i_h = 12.2f;
      static constexpr float v_l = 10.0f;
      static constexpr float v_h = 16.0f;
  };

  struct MPidConstants
  {
      // Параметры регулирования
      static constexpr float outputMin            = 0.0f;
      static constexpr float outputMaxFactor      = 1.05f;     // factor for current limit
  };

  class MStart : public MState
  {      
    public:
      MStart(MTools * Tools);
      MState * fsm() override;
  };

  class MSetCurrentMax : public MState
  {
    public:  
      MSetCurrentMax(MTools * Tools);
      MState * fsm() override;
    private:
        // Пределы регулирования max тока
      static constexpr float above = 6.0f;
      static constexpr float below = 0.2f;
  };
 
  class MSetVoltageMax : public MState
  {
    public:  
      MSetVoltageMax(MTools * Tools);
      MState * fsm() override;
  };

  class MSetCurrentMin : public MState
  {
    public:    
      MSetCurrentMin(MTools * Tools);
      MState * fsm() override;
  };

  class MSetVoltageMin : public MState
  {
    public:    
      MSetVoltageMin(MTools * Tools);
      MState * fsm() override;
  };

  class MPostpone : public MState
  {
    public:  
      MPostpone(MTools * Tools);
      MState * fsm() override;
  };

  class MUpCurrent : public MState
  {
    public:  
      MUpCurrent(MTools * Tools);
      MState * fsm() override;
  };

  class MKeepVmax : public MState
  {
    public:
      MKeepVmax(MTools * Tools);
      MState * fsm() override;
  };

  class MKeepVmin : public MState
  {
    public:  
      MKeepVmin(MTools * Tools);
      MState * fsm() override;
  };

  class MStop : public MState
  {
    public: 
      MStop(MTools * Tools);
      MState * fsm() override;
  };

  class MExit : public MState
  {
    public:
      MExit(MTools * Tools);
      MState * fsm() override;
  };

};

#endif  // !_CCCVFSM_H_
```
Не трудно заметить, что каждое состояние (далее по тексту может именоваться как "шаг") представлено соответствующим классом, а методы единственной функцией fsm() - виртуальной. 

history.txt
.LOG


20:27 12.11.2022
Новый проект Klon с клавиатурой на резисторах
Отменена емкостная клавиатура
Не тестировалось
RAM:   [=         ]  13.1% (used 42944 bytes from 327680 bytes)
Flash: [==        ]  16.1% (used 1053178 bytes from 6553600 bytes)

06:25 14.11.2022
VS, PIO new
F:\2022\Pio\Klon2
Espressif 32 3.5.0 - OK
Espressif 32 4.4.0 - error: 'vTaskEnterCritical'
DS18B20 резервируется

Проект MKlon2
Espressif 32 3.5.0
RAM:   [=         ]  13.1% (used 42944 bytes from 327680 bytes)
Flash: [==        ]  16.1% (used 1052898 bytes from 6553600 bytes)

20:12 15.11.2022
Публичные библиотеки, актуальные версии для  
  lib_deps =
  hieromon/AutoConnect@^1.3.7
  hieromon/PageBuilder@^1.5.3

vTaskEnterCritical заменен на vPortEnterCritical  (main, dallas) - не проверено
  Espressif 32    5.2.0   - OK
Dallas можно вернуть
RAM:   [=         ]  13.3% (used 43512 bytes from 327680 bytes)
Flash: [==        ]  17.5% (used 1147973 bytes from 6553600 bytes)

08:20 30.11.2022
Ошибка в топологии платы mklon2v5: IO19 освободить для MISO, RES дисплеев соединить с EN.

Определена директория для локальных библиотек
lib_extra_dirs = E:\Projects\ESP32\ESP32Lib   - OK
Перенесены ArduinoJson_ID64, AutoConnect, AutoPID_ID1925, PageBuilder_ID2010, TFT_eSPI. - OK
Библиотека дисплея TFT_eSPI используется в нескольких проектах со своими настройками, для 
этого в SP32Lib открыта директория TFT_eSPI_Setups с пользовательскими файлами xxxx_setup.h. 
После обновления библиотеки  в ней необходимо отредактировать файл User_Setup_Select.h, чтобы он 
указывал на ваш пользовательский установочный файл, например:
  #include <../TFT_eSPI_Setups/mklon25_setup.h>
Важно: вызывается только один установочный файл.
... заменить ссылку на строку Setup.h в User_Setup_Select.h на:
  #include <../TFT_eSPI_Setups/mklon25_setup.h>
и этот файл не будет перезаписан во время обновления.
Рекомендуется создать ярлык на User_Setup_Select.h

TFT_eSPI 2.3.59 обновлена до версии 2.4.79  (04.11.2022)    - OK

10:00 02.12.2022
Изменилась директория проекта E:\Projects\ESP32\MKlon2v5


src/main.cpp
/*
  project:      MESP32v7 -> Mklon -> MKlon2
  pcb:          mesp_v7.1         -> mklon2v5
  pcb:          mcdm.v7.3
  display:      1.8 дюймовый TFT ЖК-дисплей 128*160 полноцветный экран IPS. Driver IC: ST7735
  driver:       ATSAMD21 M0 MINI
  date:         2022 ноябрь
  VS:           1.73.1
  //Espressif 32: 3.5.0 (с 4.0 не совместимо)
  Espressif 32: 5.2.0 - проверить
*/

#include "board/mboard.h"
#include "board/msupervisor.h"
#include "driver/mcommands.h"
#include "display/mdisplay.h"
#include "mtools.h"
#include "mdispatcher.h"
#include "mconnmng.h"
//#include "board/mmeasure.h"
#include "measure/mmeasure.h"
//#include "dallas/temp.h"
#include "connect/connectfsm.h"
#include <Arduino.h>
  #include "project_config.h"

portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;

static MBoard      * Board      = 0;
static MDisplay    * Display    = 0;
static MTools      * Tools      = 0;
static MCommands   * Commands   = 0;
static MMeasure    * Measure    = 0;
static MDispatcher * Dispatcher = 0;
static MConnect    * Connect    = 0;

void connectTask ( void * );
void displayTask ( void * );
void coolTask    ( void * );
void mainTask    ( void * );
void measureTask ( void * );
void driverTask  ( void * );
//void tempTask    ( void * );


// setup() выполняется до запуска RTOS, а потому без обязательных требований
//  по максимальному времени монопольного захвата ядра (13мс).
void setup()
{
  //initTemp();

  Display    = new MDisplay();
  Board      = new MBoard();
  Tools      = new MTools(Board, Display);
  Commands   = new MCommands(Tools);
  Measure    = new MMeasure(Tools);
  Dispatcher = new MDispatcher(Tools);
  Connect    = new MConnect(Tools);

  
  // Выделение ресурсов для каждой задачи: память, приоритет, ядро.
  // Все задачи исполняются ядром 1, ядро 0 выделено для радиочастотных задач - BT и WiFi.
  xTaskCreatePinnedToCore ( connectTask, "Connect", 10000, NULL, 1, NULL, 1 );
  xTaskCreatePinnedToCore ( mainTask,    "Main",    10000, NULL, 2, NULL, 1 );
  xTaskCreatePinnedToCore ( displayTask, "Display",  5000, NULL, 2, NULL, 1 );
  xTaskCreatePinnedToCore ( coolTask,    "Cool",     1000, NULL, 2, NULL, 1 );
  xTaskCreatePinnedToCore ( measureTask, "Measure",  5000, NULL, 2, NULL, 1 );
  xTaskCreatePinnedToCore ( driverTask,  "Driver",   5000, NULL, 3, NULL, 1 );
  //xTaskCreatePinnedToCore ( tempTask,    "Dallas",   5000, NULL, 1, NULL, 1 );

}

void loop() 
{
  if( Serial2.available() )    // В буфере приема есть принятые байты, не факт, что пакет полный
  {                            // Пока не принят весь пакет, время ожидания ограничено 5мс
    vTaskEnterCritical(&timerMux);
      Tools->setErr( Commands->dataProcessing() );
    vTaskExitCritical(&timerMux);
  }
}

// Задача подключения к WiFi сети (полностью заимствована как есть)
void connectTask( void * )
{
  while(true) {
  //unsigned long start = millis();   // Старт таймера 
  //Serial.print("*");
    Connect->run(); 
    // Период вызова задачи задается в TICK'ах, TICK по умолчанию равен 1мс.
    vTaskDelay( 10 / portTICK_PERIOD_MS );
  // Для удовлетворения любопытства о длительности выполнения задачи - раскомментировать как и таймер.
  //Serial.print("Autoconnect: Core "); Serial.print(xPortGetCoreID()); Serial.print(" Time = "); Serial.print(millis() - start); Serial.println(" mS");
  // Core 1, 2...3 mS
  }
  vTaskDelete( NULL );
}

// Задача выдачи данных на дисплей (Закомментированы автоматически передаваемые параметры)
void displayTask( void * )
{
  while(true)
  {
    Display->runDisplay(Board->getCelsius(), Tools->getAP());
    vTaskDelay( 250 / portTICK_PERIOD_MS );
  }
  vTaskDelete( NULL );
}

// Задача управления системой теплоотвода. Предполагается расширить функциональность, добавив 
// слежение за правильностью подключения нагрузки, масштабирование тока и т.д. 
void coolTask( void * )
{
  while (true)
  {
    //unsigned long start = millis();
    Board->Supervisor->runCool();
    //Serial.print("Cool: Core "); Serial.print(xPortGetCoreID()); Serial.print(" Time = "); Serial.print(millis() - start); Serial.println(" mS");
    // Core 1, 0 mS
    vTaskDelay( 200 / portTICK_PERIOD_MS );
  }
  vTaskDelete( NULL );
}

// 1. Задача обслуживает выбор режима работы.
// 2. Управляет конечным автоматом выбранного режима вплоть да выхода из режима.
// И то и другое построены как конечные автоматы (FSM).
void mainTask ( void * )
{ 
  while (true)
  {
    // Выдерживается период запуска для вычисления амперчасов. Если прочие задачи исполняются в порядке 
    // очереди, то эта точно по таймеру - через 0,1с.
    portTickType xLastWakeTime = xTaskGetTickCount();   // To count the amp hours
    //unsigned long start = millis();
    Dispatcher->run(); 
    //Serial.print("Main: Core "); Serial.print(xPortGetCoreID()); Serial.print(" Time = "); Serial.print(millis() - start); Serial.println(" mS");
    // Core 1, 100...102 mS
    vTaskDelayUntil( &xLastWakeTime, 100 / portTICK_PERIOD_MS );    // период 0,1с
  }
  vTaskDelete( NULL );
}

// Задача управления измерениями напряжения источника питания и датчика температуры, 
void measureTask( void * )
{
  while (true)
  {
    //unsigned long start = micros();
    Measure->run();
    //Serial.print(" Time, uS: "); Serial.println(micros() - start);
    // Core 1, 32 260 mkS
    //vTaskDelay( 10 / portTICK_PERIOD_MS );
    vTaskDelay( 100 / portTICK_PERIOD_MS ); //на время экспериментов задача вызывается в 10 раз реже
  }
  vTaskDelete(NULL);
}

void driverTask( void * )
{
  while (true)
  {
    //unsigned long start = micros();
    vTaskEnterCritical(&timerMux);
      Commands->doCommand();
    vTaskExitCritical(&timerMux);
    //Serial.print(" Time, uS: "); Serial.println(micros() - start);
    // 310µS max
    vTaskDelay( 75 / portTICK_PERIOD_MS );
    //vTaskDelay( 100 / portTICK_PERIOD_MS );
  }
  vTaskDelete(NULL);
}

// void tempTask( void * parameter )
// {
//   portTickType xLastWakeTime;                                     // ******
//   xLastWakeTime = xTaskGetTickCount();                            // ******
//   while(1)
//   {
//     workTemp();
//     vTaskDelayUntil( &xLastWakeTime, 1000 / portTICK_PERIOD_MS ); 
//   }
//   vTaskDelete( NULL );
// }
